File: python/LCP_07_传递信息_easy.py
class Solution:
    def __init__(self):
        self.ways = 0
    def numWays(self, n: int, relation: List[List[int]], k: int) -> int:
        graph = collections.defaultdict(list)
        for r in relation:
            graph[r[0]].append(r[1])
        
        def dfs(index, step):
            if step == k:
                if index == n-1:
                    self.ways += 1
                return
            for to in graph[index]:
                dfs(to, step+1)
        
        dfs(0, 0)
        return self.ways

                    

File: cpp/LCP_07_传递信息_easy.cpp
class Solution {
private:
    vector<vector<int>> graph;
    int ways = 0;
    int _n;
    int _k;
public:
    int numWays(int n, vector<vector<int>>& relation, int k) {
        graph.resize(n);
        for (auto r: relation)
            graph[r[0]].push_back(r[1]);
        _n = n;
        _k = k;
        dfs(0, 0);
        return ways;
    }
    void dfs(int index, int step) {
        if (step == _k) {
            if (index == _n-1)
                ways++;
            return;
        }
        for (auto to : graph[index])
            dfs(to, step+1);
    }
};